source/slang/slang-ir-inst-pass-base.h

```diff
@@ -0,0 +1,83 @@
+// slang-ir-inst-pass-base.h
+#pragma once
+
+#include "slang-ir.h"
+#include "slang-ir-insts.h"
+
+namespace Slang
+{
+    struct IRModule;
+
+    class InstPassBase
+    {
+    protected:
+        IRModule* module;
+        SharedIRBuilder sharedBuilderStorage;
+
+        List<IRInst*> workList;
+        HashSet<IRInst*> workListSet;
+        void addToWorkList(IRInst* inst)
+        {
+            if (workListSet.Contains(inst))
+                return;
+
+            workList.add(inst);
+            workListSet.Add(inst);
+        }
+
+    public:
+        InstPassBase(IRModule* inModule)
+            : module(inModule)
+        {}
+
+        template <typename InstType, typename Func>
```

Comment:
I guess this could pull the casting 'isAImpl' function from the InstType, and so could work for non leaf types. Would also not need to pass in the inst Op or assume it matched. Could also just use static_cast

Reply:
Good point.